libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,555 @@
+      {
+      QMap<int, int> notes;
+      int rootPitch = tpc2pitch(rootTpc) + transposeOffset;
+      //euclidian mod, we need to treat this new pitch as a pitch between
+      //0 and 11, so that voicing remains consistent across transposition
+      if (rootPitch < 0)
+            rootPitch += PITCH_DELTA_OCTAVE;
+      else
+            rootPitch %= PITCH_DELTA_OCTAVE;
+
+      //create root note or bass note in second octave below middle C
+      if (bassTpc != Tpc::TPC_INVALID && voicing != Voicing::ROOT_ONLY)
+            notes.insert(tpc2pitch(bassTpc) + transposeOffset
+                        + 3*PITCH_DELTA_OCTAVE, bassTpc);
+      else
+            notes.insert(rootPitch + 3*PITCH_DELTA_OCTAVE, rootTpc);
+
+
+      switch (voicing) {
+            case Voicing::ROOT_ONLY: //already added root/bass so we are good
+                  break;
+            case Voicing::AUTO: //auto is close voicing for now since it is the most robust
+            case Voicing::CLOSE://Voices notes in close position in the first octave above middle C
+                  // FALLTHROUGH
+                  {
+                  notes.insert(rootPitch + 5*PITCH_DELTA_OCTAVE, rootTpc);
+                  //ensure that notes fall under a specific range
+                  //for now this range is between 5*12 and 6*12
+                  QMap<int, int> intervals = getIntervals(rootTpc, literal);
+                  QMapIterator<int, int> i(intervals);
+                  while (i.hasNext()) {
+                        i.next();
+                        notes.insert((rootPitch + (i.key() % 128)) % PITCH_DELTA_OCTAVE +
+                                      5*PITCH_DELTA_OCTAVE, i.value());
+                        }
+                  }
+                  break;
+            case Voicing::DROP_2:
+                  {
+                  //select 4 notes from list
+                  QMap<int, int> intervals = normalizeNoteMap(getIntervals(rootTpc, literal), rootTpc, rootPitch, 4);
+                  QMapIterator<int, int> i(intervals);
+                  i.toBack();
+
+                  int counter = 0; //counter to drop the second note
+                  while (i.hasPrevious()) {
+                        i.previous();
+                        if (++counter == 2)
+                              notes.insert(i.key() + 4*PITCH_DELTA_OCTAVE, i.value());
+                        else
+                              notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
+                        }
+                  }
+                  break;
+            case Voicing::THREE_NOTE:
+                  {
+                  //Insert 2 notes in the octave above middle C
+                  QMap<int, int> intervals = normalizeNoteMap(getIntervals(rootTpc, literal), rootTpc, rootPitch, 2, true);
+                  QMapIterator<int, int> i(intervals);
+
+                  i.next();
+                  notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
```

Comment:
Might as well use the default octave constant throughout this code.  It could eventually be made configurable, and be an instrument property (eg, for ukulele chords)

Reply:
That would be good. I considered creating properties for bass and such to allow for more flexibility.  Having the instrument property stuff would definitely be nice, although it might be a bit of work to handle it all.